### PR TITLE
Generate integrity hash values for all 'JS' files.

### PR DESCRIPTION
**Pull request by murthys**
This webpack plugin will generate the Integrity hash values run time for all the assets (js files).
This will produce two files under 'dist' folder: assets-Manifest.json, Index.html: we need to use the Json to provision the values wherever required.

This change also includes:
- Removing version folder under NPM an CDN artifacts

### CODE REVIEW

**Review by MarcinJuraszek — CHANGES_REQUESTED** [review]
This is targetting the wrong branch. It should be against 2.0-preview, not master.

**Comment by MarcinJuraszek** [comment]
hmm, looks like I was about to change the target branch :D I had no idea I have so much power in this repo! 📦

**Comment by erinha** [comment]
<p>Analyzed commit id : feb528ff7d273a2792744db8cd2d256fa0516bf1</p><hr><details><summary><b><span style="color: green">⯆</span> @microsoft/bundle-analysis-app:</b> -7 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>59.19 KB</td><td>59.19 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr><tr><td>Total Size</td><td>59.19 KB</td><td>59.19 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr></table></details><hr><p>Baseline commit: 7e2c9b9a1f4c3dfee32b1decddcb731e42fc1907</p>

**Comment by erinha** [comment]
<p>Analyzed commit id : 59a6f8807f1687ea0564c85de94303440e0efbf4</p><hr><details><summary><b><span style="color: green">⯆</span> @microsoft/bundle-analysis-app:</b> -7 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>59.19 KB</td><td>59.19 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr><tr><td>Total Size</td><td>59.19 KB</td><td>59.19 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr></table></details><hr><p>Baseline commit: 7e2c9b9a1f4c3dfee32b1decddcb731e42fc1907</p>

**Comment by erinha** [comment]
<p>Analyzed commit id : 2cccd0cf62ed552562051e7dc2f2c27221e2dfad</p><hr><details><summary><b><span style="color: green">⯆</span> @microsoft/bundle-analysis-app:</b> -7 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>59.19 KB</td><td>59.19 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr><tr><td>Total Size</td><td>59.19 KB</td><td>59.19 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr></table></details><hr><p>Baseline commit: 7e2c9b9a1f4c3dfee32b1decddcb731e42fc1907</p>

**Comment by erinha** [comment]
<p>Analyzed commit id : 1a40735a2325afb38a0726c9265d6caf2e0a5398</p><hr><details><summary><b><span style="color: green">⯆</span> @microsoft/bundle-analysis-app:</b> -7 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>59.19 KB</td><td>59.19 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr><tr><td>Total Size</td><td>59.19 KB</td><td>59.19 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr></table></details><hr><p>Baseline commit: 7e2c9b9a1f4c3dfee32b1decddcb731e42fc1907</p>

**Comment by erinha** [comment]
<p>Analyzed commit id : ab6b2e3e331deb2c69214457fc07de905ab324af</p><hr><details><summary><b><span style="color: green">⯆</span> @microsoft/bundle-analysis-app:</b> -7 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>59.19 KB</td><td>59.19 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr><tr><td>Total Size</td><td>59.19 KB</td><td>59.19 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr></table></details><hr><p>Baseline commit: 7e2c9b9a1f4c3dfee32b1decddcb731e42fc1907</p>

**Comment by erinha** [comment]
$(bundleAnalysisTask.bundleAnalysisComment)

File: azure-pipelines.yml [hunk 1]
```diff
@@ -209,31 +209,31 @@ steps:
       package: '$(Build.ArtifactStagingDirectory)\teams-test-app\$(Build.BuildId).zip'
     displayName: 'Deploy to cloudroll'
     
-  # TODO - delete after release pipeline creation
```
Comment (erinha) [inline azure-pipelines.yml:212]: We talked about this briefly yesterday in the sync. After seeing this PR in more detail, I think the changes in this file (azure-pipelines.yml) should be moved into a separate PR please.

File: packages/teams-js/webpack.config.js [hunk 2]
```diff
@@ -52,5 +62,19 @@ module.exports = {
...
+    new HtmlWebpackPlugin(),
+    new SubresourceIntegrityPlugin({
+      enabled: true,
+    }),
+    new WebpackAssetsManifest({ integrity: true }),
+    {
+      apply: compiler => {
+        compiler.hooks.done.tap('wsi-test', () => {
+          const manifest = JSON.parse(readFileSync(join(__dirname, 'dist/assets-manifest.json'), 'utf-8'));
+          expect(manifest['MicrosoftTeams.js'].integrity).toMatch(/sha256-.*/);
```
Comment (erinha) [inline packages/teams-js/webpack.config.js:74]: Is there a reason why we're using "sha256" here? To match what Teams v1 does, we should be generating sha384 hashes. I think I've seen in other discussions with you that this might not be a problem here, but I just wanted to call it out in case there is an impact here.

**Comment by erinha** [comment]
$(bundleAnalysisTask.bundleAnalysisComment)

File: packages/teams-js/webpack.config.js [hunk 3]
```diff
@@ -52,5 +64,34 @@ module.exports = {
...
+          expect(manifest['MicrosoftTeams.min.js'].integrity).toMatch(/sha384-.*/);
+
+          // Updating the old hash value with the new one in all places.
+          let currentHashValue = 'sha384-RhShpWwRxSnc+keX0WZxCsf8olaKOOw416Jky+StW6qhctsNR3GMwav/hMB1Snas';
+          replace({
+            regex: currentHashValue,
+            replacement: manifest['MicrosoftTeams.min.js'].integrity,
+            paths: ['.'],
+            recursive: true,
+            silent: true,
```
Comment (erinha) [inline packages/teams-js/webpack.config.js:91]: I don't think this is exactly what we want. All places we have HTML script tags with URLs and integrity hashes should always have a tuple that match, e.g., we don't want to have the prod CDN URL in our README file HTML script tag matched with an integrity hash that won't work with it.

File: packages/teams-js/webpack.config.js [hunk 4]
```diff
@@ -52,5 +64,34 @@ module.exports = {
...
       outputAsModuleFolder: true,
     }),
+
+    // https://www.npmjs.com/package/webpack-subresource-integrity
+    new SubresourceIntegrityPlugin({ enabled: true }),
+
+    // Webpackmanifest produces the json file containing asset(JS file) and its corresponding hash values(Example: https://github.com/waysact/webpack-subresource-integrity/blob/main/examples/webpack-assets-manifest/webpack.config.js)
+    new WebpackAssetsManifest({
+      integrity: true,
+      integrityHashes: ['sha384'],
```
Comment (erinha) [inline packages/teams-js/webpack.config.js:74]: With this change are you still generating all the different hashes (e.g., sha256, sha384, sha512, etc.)? Or is it only sha384 now?

Comment (murthys) [reply packages/teams-js/webpack.config.js:73]: No it will generate only sha384

Comment (erinha) [reply packages/teams-js/webpack.config.js:73]: Okay thanks

**Review by erinha — APPROVED** [review]
LGTM, thanks!

**Comment by erinha** [comment]
$(bundleAnalysisTask.bundleAnalysisComment)
